jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/upload/EditLambdaDialog.kt

```diff
@@ -0,0 +1,363 @@
+// Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package software.aws.toolkits.jetbrains.services.lambda.upload
+
+import com.intellij.openapi.application.ApplicationManager
+import com.intellij.openapi.application.ModalityState
+import com.intellij.openapi.application.runInEdt
+import com.intellij.openapi.module.ModuleUtil
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.ui.DialogWrapper
+import com.intellij.openapi.ui.Messages
+import com.intellij.openapi.ui.ValidationInfo
+import com.intellij.util.text.nullize
+import icons.AwsIcons
+import org.intellij.lang.annotations.Language
+import software.amazon.awssdk.services.iam.IamClient
+import software.amazon.awssdk.services.lambda.LambdaClient
+import software.amazon.awssdk.services.lambda.model.Runtime
+import software.amazon.awssdk.services.s3.S3Client
+import software.aws.toolkits.core.s3.regionForBucket
+import software.aws.toolkits.jetbrains.core.AwsClientManager
+import software.aws.toolkits.jetbrains.core.credentials.ProjectAccountSettingsManager
+import software.aws.toolkits.jetbrains.services.iam.CreateIamRoleDialog
+import software.aws.toolkits.jetbrains.services.iam.IamRole
+import software.aws.toolkits.jetbrains.services.iam.listRolesFilter
+import software.aws.toolkits.jetbrains.services.lambda.Lambda.findPsiElementsForHandler
+import software.aws.toolkits.jetbrains.services.lambda.LambdaFunction
+import software.aws.toolkits.jetbrains.services.lambda.LambdaPackager
+import software.aws.toolkits.jetbrains.services.lambda.runtimeGroup
+import software.aws.toolkits.jetbrains.utils.notifyError
+import software.aws.toolkits.jetbrains.utils.notifyInfo
+import software.aws.toolkits.jetbrains.utils.ui.addAndSelectValue
+import software.aws.toolkits.jetbrains.utils.ui.blankAsNull
+import software.aws.toolkits.jetbrains.utils.ui.populateValues
+import software.aws.toolkits.jetbrains.utils.ui.selected
+import software.aws.toolkits.resources.message
+import java.awt.event.ActionEvent
+import java.util.concurrent.TimeUnit
+import java.util.function.Function
+import javax.swing.Action
+import javax.swing.JComponent
+
+private val DEFAULT_TIMEOUT = TimeUnit.MINUTES.toSeconds(1).toInt()
+private val MAX_TIMEOUT = TimeUnit.MINUTES.toSeconds(15).toInt()
+private const val MIN_TIMEOUT = 1
+
+class EditLambdaDialog(
+    private val project: Project,
+    private val isUpdate: Boolean,
+    name: String = "",
+    description: String = "",
+    runtime: Runtime? = null,
+    handlerName: String = "",
+    envVariables: Map<String, String> = emptyMap(),
+    timeout: Int = DEFAULT_TIMEOUT,
+    role: IamRole? = null
+) : DialogWrapper(project) {
+
+    constructor(project: Project, lambdaFunction: LambdaFunction) :
```

Comment:
I'd rather push this to a static factory method but I guess it's not a big deal.